### ISSUE TITLE
Possible bug in the mapper `symbolize_keys`

### ISSUE THREAD

**Issue by nepalez** [rom-rb/rom#265]
Possible bug in the mapper `symbolize_keys`

I think this operation by its very nature should symbolize output keys whatever keys (strings or symbols) are in data.

Now it does something different [treating all the source keys as strings](https://github.com/rom-rb/rom/blob/master/lib/rom/mapper/attribute_dsl.rb#L319-L321).

That's why [this test](https://github.com/rom-rb/rom/blob/master/spec/integration/mappers/symbolizing_attributes_spec.rb#L46-L73) wouldn't pass if `task_priority` were a symbol:

```ruby
  it 'automatically maps all attributes using settings for wrap block' do
    module Test
      class TaskMapper < ROM::Mapper
        relation :tasks
        symbolize_keys true

        attribute :title

        wrap :details, prefix: 'task' do
          attribute :priority
          attribute :description
        end
      end
    end

    rom.relations.tasks << {
      'title' => 'Task One',
      task_priority: 1,               # <-- this change breaks the test
      'task_description' => 'It is a task'
    }

    task = rom.relation(:tasks).as(:tasks).first

    expect(task).to eql(
      title: 'Task One',
      details: { priority: 1, description: 'It is a task' }
    )
  end
```

I recognize the use case is a bit weird, but who knows...

My question is, should we do something with the feature, or maybe with the name of the operation.

**Comment by solnic** [comment]
I agree. It's misleading because what it really is right now is a shortcut so that you don't have to write `attribute :foo, from: "foo"`.

Maybe we should make it so that it becomes a preprocessing step that symbolizes everything (deeply as a option)?

**Comment by nepalez** [comment]
I think the consistent behaviour should follow 2 simple rules:

1) If the attibute is set as a string it takes key from string (`:from` setting by default)

All the examples below should be synonyms:

```ruby
attribute 'user_name'
```

```ruby
attribute 'user_name', from: 'user_name'
```

```ruby
prefix 'name'
attribute 'user'
```

```ruby
prefix :name
attribute 'user'
```

In the last case ^ the attribute is a string, so its source is a string too whatever prefix is.

Mutatis mutandis for symbols.

2) If the `symbolize_keys` is set, it should convert source keys to SYMBOLS instead of strings. This is a backward-incompatible change.

This should be ok

```ruby
symbolize_keys
attribute :user_name # from: :user_name
```

This shouldn't return the name because after `symbolize_keys` the `'user_name'` key is absent, only `:user_name` is.

```ruby
symbolize_keys
attribute 'user_name' # from: 'user_name'
```

Mutatis mutandis for `stringify_keys`.

---

As a bonus, this makes all 3 methods: `reject_keys`, `symbolize_keys`, `stringify_keys` being referred to source keys.
For now "symbolize" in `symbolize_keys` refers to output that is confusing.

As the next bonus we could do `symbolize_keys` and `stringify_keys` working in the same manner as `prefix` and `prefix_separator`. Namely, they could affect only on the following attributes in a current scope:

```ruby
class UserMapper < ROM::Mapper
  symbolize_keys

  wrap :contact do
    stringify_keys

    attribute 'email'
    attribute 'skype'    
  end

  attribute :id
  attribute :name
end

users.to_a
# [{ 'id' => 1, 'name' => 'jane', email: 'jane@doe.com', skype: 'jane' }]
users.as(:user).to_a
# [{ contact: { 'email' => 'jane@doe.com', skype => 'jane' }, id: 1, name: 'jane' }]
```

Why such a weird case? Because:

* this makes DSL following the principle of least surprise
* who knows what a user will want to do with his/her/its data: "the night is dark and full of horror".

**Comment by nepalez** [comment]
I don't think we should include keys transformation to special preprocessing. I like the current mechanism because it allow multiple switching from `stringify_keys` to `symbolize_keys` "and back again" when necessary (see the latest example above).

**Comment by solnic** [comment]
I need to think this through, definitely not something doable for 0.8.0 though.

**Comment by nepalez** [comment]
I agree.

While the implementation is trivial, the deprecation needed for switching `symbolize_keys` to the opposite behaviour. It couldn't be included to 0.8 even if we wanted.

**Comment by solnic** [comment]
This issue was moved to rom-rb/rom-mapper#9
